README.md
[![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version/mailR)](http://cran.r-project.org/web/packages/mailR)
[![Downloads](https://cranlogs.r-pkg.org/badges/mailR)](http://cran.rstudio.com/package=mailR)

Overview
========
mailR allows users to send emails from R.

It is developed as a wrapper around [Apache Commons Email](http://commons.apache.org/proper/commons-email/) and offers several features to send emails from R such as:
- using authentication-based SMTP servers
- sending emails to multiple recipients (including the use of Cc, Bcc, and ReplyTo recipients)
- attaching multiple files from the file system or from URLs
- sending HTML formatted emails with inline images

Note: This is the new home for mailR. The previous repository [https://github.com/rpremraj/mailR](https://github.com/rpremraj/mailR) is discontinued.


Installation instructions
=========================
You can install the latest development version of mailR using devtools:

```R
install.packages("devtools", dep = T)
library(devtools)
install_github("rpremrajGit/mailR")

library(mailR)
```

The latest release of mailR is available on [CRAN](http://cran.r-project.org/web/packages/mailR/):

```R
install.packages("mailR", dep = T)

library(mailR)
```

Usage
=====
To send an email via a SMTP server that does not require authentication:

```R
send.mail(from = "sender@gmail.com",
          to = c("Recipient 1 <recipient1@gmail.com>", "recipient2@gmail.com"),
          cc = c("CC Recipient <cc.recipient@gmail.com>"),
          bcc = c("BCC Recipient <bcc.recipient@gmail.com>"),
          subject = "Subject of the email",
          body = "Body of the email",
          smtp = list(host.name = "aspmx.l.google.com", port = 25),
          authenticate = FALSE,
          send = TRUE)
```

*Note that aspmx.l.google.com works for gmail recipients only. Check your gmail spam folder if using this server.*

To send an email via a SMTP server that requires authentication:

```R
send.mail(from = "sender@gmail.com",
          to = c("recipient1@gmail.com", "Recipient 2 <recipient2@gmail.com>"),
          replyTo = c("Reply to someone else <someone.else@gmail.com>")
          subject = "Subject of the email",
          body = "Body of the email",
          smtp = list(host.name = "smtp.gmail.com", port = 465, user.name = "gmail_username", passwd = "password", ssl = TRUE),
          authenticate = TRUE,
          send = TRUE)
```

To send an email with utf-8 or other encoding:

```R
email <- send.mail(from = "Sender Name <sender@gmail.com>",
                   to = "recipient@gmail.com",
                   subject = "A quote from Gandhi",
                   body = "In Hindi :  थोडा सा अभ्यास बहुत सारे उपदेशों से बेहतर है।
                   English translation: An ounce of practice is worth more than tons of preaching.",
                   encoding = "utf-8",
                   smtp = list(host.name = "smtp.gmail.com", port = 465, user.name = "gmail_username", passwd = "password", ssl = T),
  			   authenticate = TRUE,
				   send = TRUE)
```           
           
To send an email with one or more file attachments, and set the debug parameter to see a detailed log message:

```R
send.mail(from = "sender@gmail.com",
          to = c("recipient1@gmail.com", "recipient2@gmail.com"),
          subject = "Subject of the email",
          body = "Body of the email",
          smtp = list(host.name = "smtp.gmail.com", port = 465, user.name = "gmail_username", passwd = "password", ssl = TRUE),
          authenticate = TRUE,
          send = TRUE,
          attach.files = c("./download.log", "upload.log", "https://dl.dropboxusercontent.com/u/5031586/How%20to%20use%20the%20Public%20folder.rtf"),
          file.names = c("Download log.log", "Upload log.log", "DropBox File.rtf"), # optional parameter
          file.descriptions = c("Description for download log", "Description for upload log", "DropBox File"), # optional parameter
          debug = TRUE)
```

To send a HTML formatted email:

```R
send.mail(from = "sender@gmail.com",
          to = c("recipient1@gmail.com", "recipient2@gmail.com"),
          subject = "Subject of the email",
          body = "<html>The apache logo - <img src=\"http://www.apache.org/images/asf_logo_wide.gif\"></html>", # can also point to local file (see next example)
          html = TRUE,
          smtp = list(host.name = "smtp.gmail.com", port = 465, user.name = "gmail_username", passwd = "password", ssl = TRUE),
          authenticate = TRUE,
          send = TRUE)
```

To send a HTML formatted email with embedded inline images:

```R
send.mail(from = "sender@gmail.com",
          to = c("recipient1@gmail.com", "recipient2@gmail.com"),
          subject = "Subject of the email",
          body = "path.to.local.html.file",
          html = TRUE,
          inline = TRUE,
          smtp = list(host.name = "smtp.gmail.com", port = 465, user.name = "gmail_username", passwd = "password", ssl = TRUE),
          authenticate = TRUE,
          send = TRUE)
```
*send.mail expects the images in the HTML file to be referenced relative to the current working directory (something to improve upon in the future).*

You can add headers to your emails by passing a named list called `headers`. Some mail clients allow defining rules based on email headers, where such a feature can come in handy.

```R
send.mail(from = "sender@gmail.com",
          to = c("Recipient 1 <recipient1@gmail.com>", "recipient2@gmail.com"),
          subject = "Subject of the email",
          body = "Body of the email",
          smtp = list(host.name = "aspmx.l.google.com", port = 25),
          headers = list("X-Department" = "Finance", "X-Source" = "Automated report"),
          authenticate = FALSE,
          send = TRUE)
```

This will result in additional headers added to your email as below:
```
X-Department: Finance
X-Source: Automated report
```

MS Exchange server
==================
Provided you have the correct SMTP settings, mailR plays well with MS Exchange. Two mailR users confirmed being able to send emails via Exchange using the following code. I also successfully use mailR at work connecting to MS Exchange.
```R
send.mail(from = from,
          to = to,
          subject = subject,
          body = msg, 
          authenticate = TRUE,
          smtp = list(host.name = "smtp.office365.com", port = 587,
                      user.name = "xxx@domain.com", passwd = "xxx", tls = TRUE))
```

Sending HTML files compiled using Markdown
==========================================
mailR does not currently support resolving inline images encoded using the [data URI scheme](http://en.wikipedia.org/wiki/Data_URI_scheme). Use the workaround below instead:

First off, create the HTML file from the R terminal (the important thing here is that options does not include "base64_images" --- see `?markdown::markdownHTMLOptions`):
```R
library(knitr)
knit2html("my_report.Rmd", options = "")
```

Now you can send the resulting HTML file via mailR:
```R
send.mail(from = "sender@gmail.com",
          to = c("recipient1@gmail.com", "recipient2@gmail.com"),
          subject = "HTML file generated using Markdown",
          body = "my_report.html",
          html = TRUE,
          inline = TRUE,
          smtp = list(host.name = "smtp.gmail.com", port = 465, user.name = "gmail_username", passwd = "password", ssl = TRUE),
          authenticate = TRUE,
          send = TRUE)
```

mailR equivalent of "Blue Screen of Death"
==========================================
Many folks have run into this error using mailR: `Error in ls(envir = envir, all.names = private)`.

The source of the problem most likely lies in the connection to the SMTP server. This could either be due to incorrect settings (someone confirmed trying several servers at work until finally discovering the one playing well with mailR) or be a proxy issue. Turn on the `debug` parameter to get more pointers to resolving your problem.

While I will do my best to support you, there is little I can do remotely if you fall into this trap.

Issues/Contibutions
===================
Please post any issues encountered using mailR to the [issue tracker](https://github.com/rpremrajGit/mailR/issues).

If you would like to submit a patch to improve mailR, please send a pull request to the *develop branch*.



NEWS.md
# mailR 0.8.1 - 10 December 2021
#### What's new in version 0.8.1
- Updates to Readme
- Created News file

# mailR 0.8 - 5th November 2021
#### What's new in version 0.8
##### **Fixes **
- Added activation.jar dependency
- Switched from deprecated function setSmtpPort to setSSLOnConnect when using SSL
- Switched from deprecated function setTLS to setStartTLSEnabled when using SSL
- Internal function .valid.email() actually works now!

# mailR 0.7 - 24th August 2021
#### What's new in version 0.7 (Thank you, [sclewis23](https://github.com/sclewis23))
##### **Updates**
- Upgraded Commons Email Jar to version 1.5
- Upgraded Javax.mail Jar to version 1.6.2

# mailR 0.6 - 18th January 2016
#### What's new in version 0.6
##### **Enhancements**
- Refinement to the stripped down text version of HTML emails for incompatible clients (thanks to @dtenenba)
- Ability to add email headers by passing a named list `headers` (thanks to @dtenenba)

# mailR 0.5 - 6th December 2015
#### What's new in version 0.5
##### **Enhancements**
- Better handling of errors thrown by Java (thanks to @chlorenz)
- Email clients unable to view HTML emails now receive stripped down version of message (Fixes #24)

# mailR 0.4 - 30th December 2014
#### What's new in version 0.4
##### **Features**
- Attach files to the email using URLs, e.g., you can send files from your Dropbox public folder using the URL.
- A 'debug' parameter to set that will make send.mail() provide a detailed log.
- Option to set a email address to reply to using the 'replyTo' parameter.

#####  **Enhancement**
- Upgraded Commons Email Jar to version 1.3.3
- Upgraded Javax.mail Jar to version 1.5.2

# mailR 0.3.1 - 08th September 2014
#### What's new in version 0.3.1
##### **Enhancement**
- Better resolution of paths to allow attaching files from locations other than the working directory.
- Updated documentation to give example of use on MS Exchange

# mailR 0.3 - 12th May 2014
#### What's new in version 0.3
##### **Features**
- Added support to encode emails using iso-8859-1, utf-8, us-ascii, and koi8-r character sets.
- The body parameter can point to a locally stored text (or HTML) file and mailR will parse its contents to create the body of the email.

##### **Bug fixes**
- Experimental: changed called methods to set SSL/TLS to true to check whether it resolves issue that causes port number to default to 465.

# mailR 0.2 - 20th April 2014
#### What's new in version 0.2
##### **Features**
- mailR now allows sending email content as HTML including allowing for embedding images as inline (currently an experimental feature).
- Email addresses conforming to RFC 2822 allowed, e.g., "FirstName LastName <sender@domain.com>" allowed.
- A java stacktrace is printed out in case of failure when sending the email to allow better root cause analysis.

##### **Bug fixes**
- Fixed a bug that incorrectly set the TLS parameter as TRUE whenever the SSL parameter was set as TRUE.


R/mailR.R
#' Internal function to create file attachment objects
#'
#' @keywords internal
#' @noRd
#' @param attach.files A character vector of paths to files in the file system or valid URLs to be attached.
#' @param dots A list generated from the ellipsis parameters in send.mail. See details for more info.
#' @return attachments A vector of Java objects of class org.apache.commons.mail.EmailAttachment
#' @details The relevant optional parameters include 'file.names' listing names to assign to the attached files and 'file.descriptions' that are descriptions to be assigned to the files. If included, both paramters should be of the same length as
.createEmailAttachments <- function(attach.files, dots = NULL)
{
  file.names <- NULL
  file.descriptions <- NULL

  if(!is.null(dots))
  {
    if("file.names" %in% names(dots))
      file.names <- dots$file.names

    if("file.descriptions" %in% names(dots))
      file.descriptions <- dots$file.descriptions
  }

  if(!is.null(file.names) & length(file.names) != length(attach.files))
    stop("If not NULL, length of argument 'file.names' must equal length of argument 'file.paths'")

  if(!is.null(file.descriptions) & length(file.descriptions) != length(attach.files))
    stop("If not NULL, Length of argument 'file.descriptions' must equal length of argument 'file.paths'")

  attachments <- list()

  for(i in 1:length(attach.files))
  {
    attachments[[i]] <- .jnew("org.apache.commons.mail.EmailAttachment")
    if(isUrl(attach.files[i]))
      attachments[[i]]$setURL(.jnew("java.net.URL", attach.files[i]))
    else
    {
      if(file.exists(attach.files[i]))
        attachments[[i]]$setPath(normalizePath(attach.files[i]))
      else
        stop("Argument 'attach.files' must link to valid files")
    }

    if(!is.null(file.names)) attachments[[i]]$setName(file.names[i])
    if(!is.null(file.descriptions)) attachments[[i]]$setDescription(file.descriptions[i])
  }

  return(attachments)
}

#' Internal function to embed referenced images as inline
#'
#' @keywords internal
#' @noRd
#' @param image.file.locations A vector of paths to images files in the file system to be embedded
#' @return file.resolver A vector of Java objects of class org.apache.commons.mail.resolver.DataSourceFileResolver
#' @details This is an internal function that performs the groundwork to embed images as inline.
.resolveInlineImages <- function(image.file.locations)
{
  base_dir <- .jnew("java.io.File", normalizePath(getwd()))
  file.resolver <- .jnew("org.apache.commons.mail.resolver.DataSourceFileResolver", base_dir)
  sapply(image.file.locations, file.resolver$resolve)

  return(file.resolver)
}

#' Internal function to establish authenticated connection with SMTP server
#'
#' @keywords internal
#' @noRd
#' @param smtp A list of parameters to establish and authorize a connection with the SMTP server. See details for the various parameters.
#' @return smtp.authentication A Java object of class 'org.apache.commons.mail.DefaultAuthenticator'
.authenticateSMTP <- function(smtp)
{
  if(!all(c("user.name", "passwd") %in% names(smtp)))
    stop("Username and password required for SMTP authentication.")

  smtp.authentication <- .jnew("org.apache.commons.mail.DefaultAuthenticator", smtp$user.name, smtp$passwd)

  return(smtp.authentication)
}

#' Internal function to strip HTML tags from a string
#'
#' @keywords internal
#' @noRd
#' @param htmlString String to strip HTML tags from.
#' @return string
.stripHTML <- function(htmlString) {
  return(gsub("<.*?>", "", htmlString))
}


#' Internal function to set encoding of the email
#'
#' @keywords internal
#' @noRd
#' @param email Commons email object
#' @param encoding Character encoding to use for the email. Supported encodings include iso-8859-1 (default), utf-8, us-ascii, and koi8-r.
#' @return email Commons email object with set encoding
.resolveEncoding <- function(email, encoding)
{
  switch(encoding,
         "iso-8859-1" = {email$setCharset("iso-8859-1")},
         "utf-8" = {email$setCharset("utf-8")},
         "us-ascii" = {email$setCharset("us-ascii")},
         "koi8-r" = {email$setCharset("koi8-r")},
         stop("Supported encodings include iso-8859-1, utf-8, us-ascii, and koi8-r only.")
  )
  return(email)
}

#' Send emails from R
#'
#' @param from A valid email address of the sender.
#' @param to A character vector of recipient valid email addresses.
#' @param subject Subject of the email.
#' @param body Body of the email as text. If the parameter body refers to an existing file location, the text of the file is parsed as body of the email.
#' @param encoding Character encoding to use for the email. Supported encodings include iso-8859-1 (default), utf-8, us-ascii, and koi8-r.
#' @param html A boolean indicating whether the body of the email should be parsed as HTML.
#' @param inline A boolean indicating whether images in the HTML file should be embedded inline.
#' @param smtp A list of configuration parameters to establish and authorize a connection with the SMTP server. See details for the various parameters.
#' @param authenticate A boolean variable to indicate whether authorization is required to connect to the SMTP server. If set to true, see details on parameters required in smtp parameter.
#' @param timeout An integer with the number of milliseconds for timeouts when connecting to the SMTP server.  Default is 60s (60000ms).
#' @param send A boolean indicating whether the email should be sent at the end of the function (default behaviour). If set to false, function returns the email object to the parent environment.
#' @param attach.files A character vector of paths in the file system linking to files or *valid* URLs to be attached to the email (see details for more info on attaching URLs)
#' @param debug A boolean indicating whether you wish to see detailed debug info
#' @param ... Optional arguments to be passed related to file attachments. See details for more info.
#' @return email A Java object of class org.apache.commons.mail.SimpleEmail or org.apache.commons.mail.MultiPartEmail
#' @details The only mandatory value in the list 'smtp' is host.name that is the SMTP server address. A port number can also be provided via the list item 'port'. In case the SMTP server requires authorization, the parameter 'authenticate' must be set to TRUE and the list 'smtp' must include items 'user.name' and 'passwd'. If SSL or TLS encryption is required by the SMTP server, these can be indicated by setting a list item 'ssl' as TRUE or 'tls' as TRUE respectively.  'socketConnectionTimeout' and 'socketTimeout' allow overriding of the default 60000ms timeouts.
#'
#' Using 'attach.files' you can attach files or webpages hosted on the web (for e.g. on Dropbox). Currently, URLs hostnames must be prepending with http:// or https://. Two optional paramters relevant to attachments can be supplied. Parameter 'file.names' can be provided to assign names to the files listed in the parameter 'attach.files'. A description can be provided further as 'file.descriptions' to further describe the file. Both parameters must have the same length as 'attach.files'. In case attach.file is NULL, then these two parameters will be ignored.
#'
#' HTML formatted emails can be sent by setting the parameters html and inline (if embedding images) to TRUE. The body of the email can either be a HTML string or point to a HTML file in the local file system.
#'
#' Email headers can be set by passing a named list called headers, e.g., 'headers = list("X-Department"="Finance","X-Source"="Automated report")'.
#' @export send.mail
#' @import rJava
#' @import stringr
#' @import R.utils
#' @import assertthat
#' @note For more examples, see https://github.com/rpremraj/mailR
#' @examples
#' sender <- "sender@@gmail.com"  # Replace with a valid address
#' recipients <- c("receiver1@@gmail.com")  # Replace with one or more valid addresses
#' email <- send.mail(from = sender,
#'                    to = recipients,
#'                    subject="Subject of the email",
#'                    body = "Body of the email",
#'                    smtp = list(host.name = "aspmx.l.google.com", port = 25),
#'                    authenticate = FALSE,
#'                    send = FALSE)
#' \dontrun{email$send() # execute to send email or set parameter send= TRUE in send.mail()}
send.mail <- function(from, to, subject = "", body = "", encoding = "iso-8859-1", html = FALSE, inline = FALSE, smtp = list(), authenticate = FALSE, timeout = 60000, send = TRUE, attach.files = NULL, debug = FALSE, ...)
{
  assert_that(length(from) == 1, msg = "Argument 'from' should be a of length one.")
  assert_that(length(to) > 0, msg = "Argument 'to' must have at least one single (valid) email address.")
  assert_that("host.name" %in% names(smtp), msg = "Check documentation to include all mandatory parameters to establish SMTP connection.")

  if (authenticate)
  {
    assert_that("user.name" %in% names(smtp), msg = "List element 'user.name' missing in smtp. Check documentation to include all mandatory parameters to establish SMTP connection.")
    assert_that("passwd" %in% names(smtp), msg = "List element 'passwd' missing in smtp. Check documentation to include all mandatory parameters to establish SMTP connection.")
  }

  dots <- list(...)

  if(html && inline)
    email <- .jnew("org.apache.commons.mail.ImageHtmlEmail")
  else if(html)
    email <- .jnew("org.apache.commons.mail.HtmlEmail")
  else if(!is.null(attach.files))
    email <- .jnew("org.apache.commons.mail.MultiPartEmail")
  else
    email <- .jnew("org.apache.commons.mail.SimpleEmail")

  if(debug)
    email$setDebug(TRUE)

  email <- .resolveEncoding(email, encoding)

  if(!is.null(attach.files))
  {
    attachments <- .createEmailAttachments(attach.files, dots)
    sapply(attachments, email$attach)
  }

  if(.jclass(email) == "org.apache.commons.mail.ImageHtmlEmail")
  {
    image.files.references <- str_extract_all(body, email$REGEX_IMG_SRC)
    pattern <- "\"([^\"]*)\""
    image.files.locations <- gsub("\"", "", sapply(str_extract_all(image.files.references[[1]], pattern), "[[", 1))
    file.resolver <- .resolveInlineImages(image.files.locations)
    email$setDataSourceResolver(file.resolver)
  }

  email$setHostName(smtp$host.name)

  if("port" %in% names(smtp))
    email$setSmtpPort(as.integer(smtp$port));

  if(authenticate == TRUE)
    email$setAuthenticator(.authenticateSMTP(smtp));

  if("ssl" %in% names(smtp))
    if(smtp$ssl)
      email$setSSLOnConnect(TRUE)

  if("tls" %in% names(smtp))
    if(smtp$tls)
      email$setStartTLSEnabled(TRUE)

  if(!missing(timeout) && is.numeric(timeout)){
    email$setSocketTimeout(as.integer(timeout))
    email$setSocketConnectionTimeout(as.integer(timeout))
  }

  email$setFrom(from)
  email$setSubject(subject)

  if(file.exists(body))
    body <- readChar(body, file.info(body)$size)

  if(html)
  {
    email$setHtmlMsg(as.character(body))
    email$setTextMsg(.stripHTML(body))
  } else
    email$setMsg(body)

  if(.valid.email(to))
    sapply(to, email$addTo)

  if("cc" %in% names(dots))
  {
    if(.valid.email(dots$cc))
      sapply(dots$cc, email$addCc)
  }

  if("bcc" %in% names(dots))
  {
    if(.valid.email(dots$bcc))
      sapply(dots$bcc, email$addBcc)
  }

  if("replyTo" %in% names(dots))
  {
    if(.valid.email(dots$replyTo))
      sapply(dots$replyTo, email$addReplyTo)
  }

  if ("headers" %in% names(dots))
  {
    lapply(names(dots$headers), function(x)
      email$addHeader(x, dots$headers[[x]]))
  }

  if(send)
    .jTryCatch(email$send())

  return(email)
}

#' Internal function to validate email addresses
#'
#' @keywords internal
#' @noRd
#' @param email.addresses A character vector of email addresses expected in RFC#822 standard.
#' @return TRUE Boolean TRUE if all items in 'emails' are valid emails. If a malformed email address is identified, the function stops execution of the calling function 'send.mail' and prints the relevant item to console.
#' @examples
#' .valid.email("<user@@email.com>") # TRUE

.valid.email <- function(email.addresses)
{
  valid <- logical(length(email.addresses))

  for(i in seq_along(email.addresses))
    valid[i] <- ifelse(length(.jTryCatch(.jnew("javax.mail.internet.InternetAddress", email.addresses[i])$validate())) == 0, TRUE, FALSE)

  if(all(valid))
    return(TRUE)
}


#' Internal function to catch Java exceptions and print stack traces. Inspired by author of package XLConnect.
#'
#' @keywords internal
#' @noRd
#' @param ... A call to a Java method
.jTryCatch <- function(...) {
  tryCatch(..., Throwable =
             function(e) {
               if(!is.jnull(e$jobj)) {
                 print(e$jobj$printStackTrace())
                 stop(paste(class(e)[1], e$jobj$getMessage(), sep = " (Java): "), call. = FALSE)
               } else
                 stop("Undefined error occurred! Turn debug mode on to see more details.")
             },
             error = function(e) message(e)
  )
}
